mm2src/coins/eth/eth_balance_events.rs

```diff
@@ -16,49 +17,79 @@ use super::EthCoin;
-async fn get_all_balance_results_concurrently(
-    coin: &EthCoin,
-) -> Vec<Result<(String, BigDecimal), (String, MmError<BalanceError>)>> {
+async fn get_all_balance_results_concurrently(coin: &EthCoin) -> Vec<BalanceResult> {
     let mut tokens = coin.get_erc_tokens_infos();
 
-    // Workaround for performance purposes.
-    //
-    // Unlike tokens, the platform coin length is constant (=1). Instead of creating a generic
-    // type and mapping the platform coin and the entire token list (which can grow at any time), we map
-    // the platform coin to Erc20TokenInfo so that we can use the token list right away without
-    // additional mapping.
-    tokens.insert(coin.ticker.clone(), Erc20TokenInfo {
-        token_address: coin.my_address,
-        decimals: coin.decimals,
-    });
-
-    let jobs = tokens
-        .into_iter()
-        .map(|(token_ticker, info)| async move { fetch_balance(coin, token_ticker, &info).await })
-        .collect::<FuturesUnordered<_>>();
-
-    jobs.collect().await
+    let addresses = match coin.my_addresses().await {
+        Ok(addresses) => addresses,
+        Err(e) => {
+            return vec![Err((coin.ticker.clone(), e.into()))];
+        },
+    };
+
+    let mut all_jobs = FuturesUnordered::new();
+
+    for address in addresses {
+        // Workaround for performance purposes.
+        //
+        // Unlike tokens, the platform coin length is constant (=1). Instead of creating a generic
+        // type and mapping the platform coin and the entire token list (which can grow at any time), we map
+        // the platform coin to Erc20TokenInfo so that we can use the token list right away without
+        // additional mapping.
+        tokens.insert(coin.ticker.clone(), Erc20TokenInfo {
+            token_address: address,
+            decimals: coin.decimals,
+        });
+
+        let jobs = tokens.iter().map(|(token_ticker, info)| {
+            let coin = coin.clone();
+            let token_ticker = token_ticker.clone();
+            let info = info.clone();
+            async move { fetch_balance(&coin, address, token_ticker, &info).await }
+        });
+
+        all_jobs.extend(jobs);
+    }
+
+    let results: Vec<_> = all_jobs.collect().await;
+
+    let mut final_results: Vec<BalanceResult> = Vec::new();
+    for result in results {
+        match result {
+            Ok((ticker, (address, balance))) => {
+                let mut balances = HashMap::new();
+                balances.insert(address, balance);
+                final_results.push(Ok((ticker, balances)));
+            },
+            Err((ticker, e)) => final_results.push(Err((ticker, e))),
+        }
+    }
+
+    final_results
 }
 
 async fn fetch_balance(
     coin: &EthCoin,
+    address: Address,
     token_ticker: String,
     info: &Erc20TokenInfo,
-) -> Result<(String, BigDecimal), (String, MmError<BalanceError>)> {
+) -> Result<(String, (String, BigDecimal)), (String, MmError<BalanceError>)> {
```

Comment:
I think it's time to create a type for this. :smile: At the first look I had no idea what the extra String is used for.

Reply:
Will create a type for sure. I am conflicted about whether we should have the balance data be ticker-based or address-based. I guess having it ticker-based, as it is currently done, while adding the total balance across all addresses, will be the best solution. It will also align with how GUIs are more coin-based than address-based currently. However, this might change depending on how GUIs implement HD wallets.

Reply:
> I guess having it ticker-based, as it is currently done, while adding the total balance across all addresses, will be the best solution

Nevermind this, we only broadcast if a balance changes for an address so no need for total balance.

Reply:
fixed here https://github.com/KomodoPlatform/komodo-defi-framework/pull/1979/commits/86603926a2972fb94195f8693dd144e10c5bed13